File: blue/src/colors_compute/coordination.py
"""Pure conditional journal-write intentions; no transport or mutation authority."""
from ._copy import deepcopy
import math
import re

from .contract import _safe, expand, registry, state_keys

MAX_INTEGER = 9007199254740991


def _integer(value, minimum, maximum):
    return (type(value) in (int, float) and minimum <= value <= maximum
            and math.isfinite(value) and value == int(value))


def _shape(value, fields):
    return isinstance(value, dict) and set(value) == set(fields)


def _nonblank(value):
    return isinstance(value, str) and bool(value.strip())


def _role(value):
    return value is None or isinstance(value, str) and bool(re.fullmatch(r'[a-z][a-z0-9]*(-[a-z0-9]+)*', value))


def _identity(value):
    if not _shape(value, ('profile', 'provider', 'backend')) or not _safe(value['profile']):
        return False
    if not isinstance(value['provider'], str) or value['provider'] not in registry()['compute']:
        return False
    backend = value['backend']
    if not isinstance(backend, dict):
        return False
    kind = backend.get('kind')
    if kind not in ('r2', 's3') or not _shape(backend, ('kind', 'bucket', 'region', 'endpoint') if kind == 'r2' else ('kind', 'bucket', 'region')):
        return False
    if not isinstance(backend['bucket'], str) or not re.fullmatch(r'[a-z0-9][a-z0-9.-]{0,62}', backend['bucket']) or not _safe(backend['region']):
        return False
    return kind == 's3' or (backend['region'] == 'auto' and isinstance(backend['endpoint'], str)
                           and bool(re.fullmatch(r'https://[a-zA-Z0-9.-]+(:[0-9]{1,5})?/?', backend['endpoint'])))


def _topology(value):
    if not isinstance(value, list) or not value:
        return None
    normalized, total = [], 0
    for declaration in value:
        if not isinstance(declaration, dict) or set(declaration) - {'role', 'count'}:
            return None
        count = declaration.get('count', 1)
        if not _role(declaration.get('role')) or not _integer(count, 1, 1000):
            return None
        total += int(count)
        if total > 1000:
            return None
        normalized.append({'role': declaration.get('role'), 'count': int(count)})
    try:
        return expand(normalized)
    except ValueError:
        return None


def _event(value):
    if not isinstance(value, dict):
        return False
    kind = value.get('type')
    extra = {'acquire': (), 'declare': ('topology',), 'start': ('node_id', 'operation_id'),
             'complete': ('node_id', 'operation_id'), 'fail': ('node_id', 'operation_id'), 'release': ()}
    if not isinstance(kind, str) or kind not in extra or not _shape(value, ('type', 'run_id', 'write_id', 'target_etag', *extra[kind])):
        return False
    if not _safe(value['run_id']) or not _safe(value['write_id']) or not (value['target_etag'] is None or _nonblank(value['target_etag'])):
        return False
    if kind == 'declare':
        return _topology(value['topology']) is not None
    if kind in ('start', 'complete', 'fail'):
        return _safe(value['node_id']) and _safe(value['operation_id'])
    return True


def _observation(value):
    if not isinstance(value, dict):
        return False
    status = value.get('status')
    if status in ('absent', 'error'):
        return _shape(value, ('status',))
    return status == 'present' and _shape(value, ('status', 'etag', 'document')) and _nonblank(value['etag'])


def _document_v1(value):
    if not _shape(value, ('schema_version', 'identity', 'revision', 'write_id', 'lock', 'topology_declared', 'nodes')):
        return False
    if not _integer(value['schema_version'], 1, 1) or not _identity(value['identity']) or not _integer(value['revision'], 1, MAX_INTEGER) or not _safe(value['write_id']):
        return False
    lock = value['lock']
    if not _shape(lock, ('state', 'run_id')) or not ((lock['state'] == 'held' and _safe(lock['run_id'])) or (lock['state'] == 'idle' and lock['run_id'] is None)):
        return False
    nodes = value['nodes']
    declared = value['topology_declared']
    if type(declared) is not bool or not isinstance(nodes, dict) or len(nodes) > 1000 or declared != bool(nodes):
        return False
    roles, operations = {}, set()
    for node_id, node in nodes.items():
        if not _safe(node_id) or not _shape(node, ('state_key', 'role', 'index', 'phase', 'operation_id')):
            return False
        role, index, phase, operation = node['role'], node['index'], node['phase'], node['operation_id']
        if not _role(role) or not _integer(index, 0, 999):
            return False
        expected_id = f'{role}-{int(index)}' if role is not None else str(int(index))
        if node_id != expected_id or node['state_key'] != state_keys(value['identity']['profile'], [node_id])['nodes'][node_id]:
            return False
        if phase not in ('declared', 'running', 'ready', 'failed'):
            return False
        if phase == 'declared':
            if operation is not None:
                return False
        elif not _safe(operation) or operation in operations:
            return False
        else:
            operations.add(operation)
        if phase == 'running' and lock['state'] != 'held':
            return False
        roles.setdefault(role, []).append(int(index))
    if None in roles and len(roles) > 1:
        return False
    return all(sorted(indices) == list(range(len(indices))) for indices in roles.values())


def _document(value):
    if isinstance(value, dict) and value.get('schema_version') == 2:
        from .lifecycle import lifecycle_document_valid
        return lifecycle_document_valid(value)
    return _document_v1(value)


def coordination(observation, identity, event):
    """Return a new conditional-write intent; all inputs remain unchanged."""
    if isinstance(event, dict) and isinstance(event.get('type'), str) and event['type'].startswith('lifecycle/'):
        from .lifecycle import lifecycle
        return lifecycle(observation, identity, event)
    def fail(message):
        raise ValueError(message)
    if not _identity(identity):
        fail('invalid coordination identity')
    if not _event(event):
        fail('invalid coordination event')
    if not _observation(observation):
        fail('invalid coordination observation')
    if observation['status'] == 'error':
        fail('coordination read failed')
    present = observation['status'] == 'present'
    document = observation.get('document')
    if present:
        if not _document_v1(document):
            fail('invalid coordination document')
        if document['identity'] != identity:
            fail('coordination identity mismatch')
    if event['target_etag'] != (observation['etag'] if present else None):
        fail('stale coordination observation')
    if present and event['write_id'] == document['write_id']:
        fail('coordination write_id reused')
    if present and document['revision'] == MAX_INTEGER:
        fail('coordination revision exhausted')
    kind = event['type']
    if not present:
        if kind != 'acquire':
            fail('coordination object absent')
        return {'condition': {'if_none_match': '*'}, 'document': {
            'schema_version': 1, 'identity': deepcopy(identity), 'revision': 1,
            'write_id': event['write_id'], 'lock': {'state': 'held', 'run_id': event['run_id']},
            'topology_declared': False, 'nodes': {}}}
    result = deepcopy(document)
    if kind == 'acquire':
        if document['lock']['state'] == 'held':
            fail('coordination lock held')
        result['lock'] = {'state': 'held', 'run_id': event['run_id']}
    else:
        if document['lock']['state'] != 'held':
            fail('coordination lock not held')
        if document['lock']['run_id'] != event['run_id']:
            fail('coordination owner mismatch')
        if kind == 'declare':
            if document['topology_declared']:
                fail('coordination topology already declared')
            nodes = _topology(event['topology'])
            keys = state_keys(identity['profile'], [node['node_id'] for node in nodes])['nodes']
            result['nodes'] = {node['node_id']: {'state_key': keys[node['node_id']], 'role': node['role'],
                'index': node['index'], 'phase': 'declared', 'operation_id': None} for node in nodes}
            result['topology_declared'] = True
        elif kind in ('start', 'complete', 'fail'):
            if not document['topology_declared']:
                fail('coordination topology not declared')
            node_id = event['node_id']
            if node_id not in document['nodes']:
                fail('coordination node undeclared')
            node = document['nodes'][node_id]
            if kind == 'start':
                if node['phase'] != 'declared':
                    fail('coordination node not startable')
                if any(n['operation_id'] == event['operation_id'] for n in document['nodes'].values()):
                    fail('coordination operation_id reused')
                result['nodes'][node_id].update(phase='running', operation_id=event['operation_id'])
            else:
                if node['phase'] != 'running':
                    fail('coordination node not running')
                if node['operation_id'] != event['operation_id']:
                    fail('coordination operation mismatch')
                result['nodes'][node_id]['phase'] = 'ready' if kind == 'complete' else 'failed'
        elif kind == 'release':
            if any(node['phase'] == 'running' for node in document['nodes'].values()):
                fail('coordination operations outstanding')
            result['lock'] = {'state': 'idle', 'run_id': None}
    result['revision'] = int(document['revision']) + 1
    result['write_id'] = event['write_id']
    return {'condition': {'if_match': observation['etag']}, 'document': result}

File: red/src/coordination.ts
import {managedCoordination,managedDocumentValid} from './managed-journal.ts';
import {lifecycle,lifecycleDocumentValid} from './lifecycle.ts';
import {registry, expand, state_keys} from './index.ts';
type Map = Record<string, any>;
const object = (value: unknown): value is Map => value !== null && typeof value === 'object' && !Array.isArray(value);
const fields = (value: unknown, required: string[], optional: string[] = []): value is Map => object(value) && required.every(k => Object.hasOwn(value,k)) && Object.keys(value).every(k => [...required,...optional].includes(k));
const match = (value: unknown, pattern: RegExp): value is string => typeof value === 'string' && pattern.exec(value)?.[0] === value;
const safe = (value: unknown) => match(value,/^[a-zA-Z0-9][a-zA-Z0-9_-]{0,62}$/);
const nonblank = (value: unknown) => typeof value === 'string' && !!value.trim();
const integer = (value: unknown, min: number, max: number) => typeof value === 'number' && Number.isSafeInteger(value) && value >= min && value <= max;
const roleValid = (value: unknown) => value === null || match(value,/^[a-z][a-z0-9]*(-[a-z0-9]+)*$/);
export function identityValid(value: unknown): value is Map {
  if (!fields(value,['profile','provider','backend']) || !safe(value.profile) || typeof value.provider !== 'string' || !Object.hasOwn(registry.compute,value.provider)) return false;
  const backend=value.backend;
  if (!fields(backend,['kind','bucket','region'],['endpoint']) || !match(backend.bucket,/^[a-z0-9][a-z0-9.-]{0,62}$/) || !safe(backend.region)) return false;
  if ((backend.kind === 's3' || backend.kind === 'gcs')) return !Object.hasOwn(backend,'endpoint');
  return (backend.kind === 'oci' || (backend.kind === 'r2' && backend.region === 'auto')) && Object.hasOwn(backend,'endpoint') && match(backend.endpoint,/^https:\/\/[a-zA-Z0-9.-]+(:[0-9]{1,5})?\/?$/);
}
function topologyValid(value: unknown): value is Map[] {
  if (!Array.isArray(value) || !value.length || value.length > 1000) return false;
  const roles=new Set<string|null>();let total=0;
  for(const declaration of value) {
    if (!fields(declaration,[],['role','count'])) return false;
    const role=declaration.role ?? null;
    const count=Object.hasOwn(declaration,'count') ? declaration.count : 1;
    if (!roleValid(role) || roles.has(role) || (role === null && value.length !== 1) || !integer(count,1,1000)) return false;
    roles.add(role);total+=count;
    if (total>1000 || !safe(role === null ? String(count-1) : `${role}-${count-1}`)) return false;
  }
  return true;
}
function eventValid(value: unknown): value is Map {
  if (!object(value) || typeof value.type !== 'string') return false;
  const extra: Record<string,string[]>={acquire:[],declare:['topology'],start:['node_id','operation_id'],complete:['node_id','operation_id'],fail:['node_id','operation_id'],release:[]};
  if (!Object.hasOwn(extra,value.type) || !fields(value,['type','run_id','write_id','target_etag',...extra[value.type]])) return false;
  if (!safe(value.run_id) || !safe(value.write_id) || !(value.target_etag === null || nonblank(value.target_etag))) return false;
  if (value.type === 'declare') return topologyValid(value.topology);
  if (extra[value.type].includes('node_id')) return safe(value.node_id) && safe(value.operation_id);
  return true;
}
function observationValid(value: unknown): value is Map {
  if (!object(value)) return false;
  if(value.status === 'absent' || value.status === 'error') return fields(value,['status']);
  return value.status === 'present' && fields(value,['status','etag','document']) && nonblank(value.etag);
}
export function documentValid(value: unknown): value is Map {
  if(object(value)&&value.schema_version===3)return managedDocumentValid(value);
  if(object(value)&&value.schema_version===2)return lifecycleDocumentValid(value);
  if (!fields(value,['schema_version','identity','revision','write_id','lock','topology_declared','nodes']) || value.schema_version !== 1 || !identityValid(value.identity) || !integer(value.revision,1,Number.MAX_SAFE_INTEGER) || !safe(value.write_id)) return false;
  if (!fields(value.lock,['state','run_id']) || !((value.lock.state === 'held' && safe(value.lock.run_id)) || (value.lock.state === 'idle' && value.lock.run_id === null))) return false;
  if (typeof value.topology_declared !== 'boolean' || !object(value.nodes)) return false;
  const nodes=Object.entries(value.nodes);
  if (nodes.length > 1000 || (value.topology_declared ? nodes.length === 0 : nodes.length !== 0)) return false;
  const operations=new Set<string>();const roles=new Map<string|null,number[]>();
  for(const [id,node] of nodes) {
    if (!safe(id) || !fields(node,['state_key','role','index','phase','operation_id']) || !roleValid(node.role) || !integer(node.index,0,999)) return false;
    if (id !== (node.role === null ? String(node.index) : `${node.role}-${node.index}`) || node.state_key !== `${value.identity.profile}/compute/nodes/${id}.tfstate`) return false;
    if (!['declared','running','ready','failed'].includes(node.phase)) return false;
    if (node.phase === 'declared') {if(node.operation_id !== null) return false;}
    else {
      if (!safe(node.operation_id) || operations.has(node.operation_id)) return false;
      operations.add(node.operation_id);
    }
    if (node.phase === 'running' && value.lock.state !== 'held') return false;
    roles.set(node.role,[...(roles.get(node.role) ?? []),node.index]);
  }
  if(roles.has(null) && roles.size !== 1) return false;
  return [...roles.values()].every(indices => indices.sort((a,b)=>a-b).every((index,position)=>index === position));
}
export function identityEqual(a: Map,b: Map) {
  return a.profile===b.profile && a.provider===b.provider && ['kind','bucket','region','endpoint'].every(key=>a.backend[key]===b.backend[key]);
}
function fail(message: string): never {throw new Error(message);}

/** Plan one journal CAS. A returned plan does not authorize provider dispatch. */
export function coordination(observation: unknown, identity: unknown, event: unknown) {
  if(object(event)&&typeof event.type==='string'&&event.type.startsWith('managed/'))return managedCoordination(observation,identity,event);
  if(object(event)&&typeof event.type==='string'&&event.type.startsWith('lifecycle/'))return lifecycle(observation,identity,event);
  if(!identityValid(identity)) fail('invalid coordination identity');
  if(!eventValid(event)) fail('invalid coordination event');
  if(!observationValid(observation)) fail('invalid coordination observation');
  if(observation.status === 'error') fail('coordination read failed');
  const present=observation.status === 'present';
  const document=observation.document;
  if(present && (!documentValid(document)||document.schema_version!==1)) fail('invalid coordination document');
  if(present && !identityEqual(document.identity,identity)) fail('coordination identity mismatch');
  if(event.target_etag !== (present ? observation.etag : null)) fail('stale coordination observation');
  if(present && event.write_id === document.write_id) fail('coordination write_id reused');
  if(present && document.revision === Number.MAX_SAFE_INTEGER) fail('coordination revision exhausted');
  if(!present) {
    if(event.type !== 'acquire') fail('coordination object absent');
    return {condition:{if_none_match:'*'},document:{schema_version:1,identity:structuredClone(identity),revision:1,write_id:event.write_id,lock:{state:'held',run_id:event.run_id},topology_declared:false,nodes:{}}};
  }
  const next:Map=structuredClone(document);
  if(event.type === 'acquire') {
    if(next.lock.state === 'held') fail('coordination lock held');
    next.lock={state:'held',run_id:event.run_id};
  } else {
    if(next.lock.state !== 'held') fail('coordination lock not held');
    if(next.lock.run_id !== event.run_id) fail('coordination owner mismatch');
    if(event.type === 'declare') {
      if(next.topology_declared) fail('coordination topology already declared');
      const expanded=expand(event.topology);
      const keys=state_keys(identity.profile,expanded.map(node=>node.node_id));
      next.nodes=Object.fromEntries(expanded.map(node=>[node.node_id,{state_key:keys.nodes[node.node_id],role:node.role,index:node.index,phase:'declared',operation_id:null}]));
      next.topology_declared=true;
    } else if(event.type === 'release') {
      if(Object.values(next.nodes).some((node:any)=>node.phase === 'running')) fail('coordination operations outstanding');
      next.lock={state:'idle',run_id:null};
    } else {
      if(!next.topology_declared) fail('coordination topology not declared');
      if(!Object.hasOwn(next.nodes,event.node_id)) fail('coordination node undeclared');
      const node=next.nodes[event.node_id];
      if(event.type === 'start') {
        if(node.phase !== 'declared') fail('coordination node not startable');
        if(Object.values(next.nodes).some((n:any)=>n.operation_id===event.operation_id)) fail('coordination operation_id reused');
        node.phase='running';node.operation_id=event.operation_id;
      } else {
        if(node.phase !== 'running') fail('coordination node not running');
        if(node.operation_id !== event.operation_id) fail('coordination operation mismatch');
        node.phase=event.type === 'complete' ? 'ready' : 'failed';
      }
    }
  }
  next.revision++;next.write_id=event.write_id;
  return {condition:{if_match:observation.etag},document:next};
}
